app/controllers/tokens_controller.rb

```diff
@@ -0,0 +1,69 @@
+class TokensController < Doorkeeper::TokensController
+  include ErrorResponses
+
+  def create
+    authenticate_with_facebook if signing_in_with_facebook?
+    authenticate_with_credentials if singing_in_regularly?
+  rescue Doorkeeper::Errors::DoorkeeperError,
+         Doorkeeper::Errors::InvalidGrantReuse,
+         Doorkeeper::OAuth::Error,
+         Koala::Facebook::AuthenticationError,
+         ActiveRecord::RecordNotFound => e
+    render_error e
+  end
+
+  private
+
+    def signing_in_with_facebook?
+      params[:username] == "facebook"
```

Comment:
I wonder if we shouldn't be doing this a better way? Should we have an additional parameter to specify type of sign-in? Or is this something that's a vestige of the authentication library we were using client-side? Would maybe be good to open an issue around this at least so we can readdress when we get to client-side implementation.